### ISSUE TITLE
Fallback not working using different versions of the same module

### ISSUE THREAD

**Issue by floqqi** [mathcamp/pypicloud#82]
Scenario:

- Fetch `requests==2.9.1` to your own pypicloud (pypi.myserver.com)
- Run `pip install -i http://pypi.myserver.com/pypi/ requests==2.9.0` on your local machine

Result:

```
Could not find a version that satisfies the requirement requests==2.9.0 (from versions: 2.9.1)
No matching distribution found for requests==2.9.0
```

Shouldn't this request fallback to `pypi.python.org`?

**Comment by stevearc** [comment]
You have encountered one of the oddities of pip. Pip doesn't tell the server which version it's looking for; it only says "Hey, I want the package 'requests'". The server then has to send down all versions of that package, and the pip client will choose one to download.

The way `redirect` is currently implemented is that pypicloud will check if it has any packages named 'requests' and serve them if so. If not, it will return a 302 to pypi.python.org.

So you can see how this leads to a failure in your particular situation. I would recommend taking a look at the [exhaustive list of all fallback behaviors](http://pypicloud.readthedocs.org/en/latest/topics/fallback.html#fallback-detail) because I suspect that `mirror` may more closely match what you want.

If you still want to keep with `redirect` and just have it behave better, I can look into adding a setting that will make it do what you want. I can't promise that it'll get done soon, though. I'm being _figuratively hammered_ by work.

**Comment by floqqi** [comment]
Fallback behavior `mirror` might be a solution. But my intention was not to store all packages locally, since they are available at pypi.

My use case is the following:
- Store *some* pre-compiled packages, which are only available as sdist at `pypi.python.org` (like SQLAlchemy) in the internal pypicloud. (Or not even compilable without lots of effort, http://www.lfd.uci.edu/~gohlke/pythonlibs/ \*wink\*)
- If a `pip install …` requests another version (which is not available in the internal pypicloud) redirect to `pypi.python.org`.

Maybe a new fallback behavior is the solution? Some mix of `mirror` and `redirect` without storing/caching the package.

**Comment by stevearc** [comment]
Okay! This is a pretty reasonable feature request. I think it wouldn't be too much trouble to put this behind a setting. I'll start digging into it.

**Comment by floqqi** [comment]
Awesome! That'd be an unique feature and finally replace my current [pypiserver](https://pypi.python.org/pypi/pypiserver) installation.

Feel free to contact me if you need some help or a beta tester! :)

**Comment by stevearc** [comment]
Okay! I released version [0.3.11](https://pypi.python.org/pypi/pypicloud/0.3.11) which has the [pypi.always_show_upstream](http://pypicloud.readthedocs.io/en/0.3.11/topics/configuration.html#pypi-always-show-upstream) configuration option. Setting that to `True` should do what you want.

**Comment by floqqi** [comment]
Works perfectly fine using `pypi.always_show_upstream = true` :+1:
